src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -509,4 +1538,83 @@ public Object readResolve() {
         }
     }
 
+    /**
+     * A custom comparator when the object in the collection is of type {@link Pair<A,B>}.
+     * @param <A> Type of the first object being encapsulated.
+     * @param <B> ype of the second object being encapsulated.
+     * <p>
+     * The comparator orders the objects first in the descending order of first encapsulated object and then in increasing
+     * order of second encapsulated object.
+     */
+    private static final class PairComparator<A extends Comparable<? super A>, B extends Comparable<? super B>> implements Comparator<Pair<A, B>> {
+        public int compare (Pair<A, B> pair1, Pair<A, B> pair2) {
+            if (pair1.first.compareTo(pair2.first) != 0) {
+                return -1 * pair1.first.compareTo(pair2.first);
+            } else
+                return pair1.second.compareTo(pair2.second);
+        }
+    }
+
+    /**
+     * A generic class to encapsulate a pair of objects where each object can be any type.
+     *
+     * @param <A> Type of the first object.
+     * @param <B> Type of the second object.
+     */
+    private static final class Pair<A, B> {
+
+        /**
+         * First object.
+         */
+        A first;
+
+        /**
+         * Second object.
+         */
+        B second;
+
+        /**
+         * A constructor to create a {@link Pair} object containing both the objects
+         * {@link Pair#first} and {@link Pair#second} as pair.
+         *
+         * @param first  The first object conatined in the pair.
+         * @param second The second object contained in the pair.
+         */
+        Pair (A first, B second) {
+            this.first = first;
+            this.second = second;
+        }
+    }
+
+    /**
+     * Class overriding {@link XYToolTipGenerator} for generating custom tool tips and urls for flapper trend.
+     */
+    public static class XYToolTipGenerator2 implements XYToolTipGenerator, XYURLGenerator {
```

Comment:
`XYToolTipGenerator2` please use better name for this method

Reply:
Done.